### DISCUSSION TITLE
Using npm packages on my zeppos app

### DISCUSSION THREAD

**Discussion by Noxde** [zepp-health/zeppos-forum#329] [Q&A] [answered]
Im trying to use xml2js on my zepp os app but im only getting build errors
![image](https://github.com/zepp-health/zeppos-forum/assets/70549780/560f6cc9-8f69-4a02-9be1-9c10b2efa92e)
That is the result of doing 
```js
import parseString from "xml2js";
```
if i instead use 
```js
const { parseString } = require("xml2js");
```
I don't get the errors but it doesn't work either, I'm guessing i can't use any npm packages that I install that are not from @zeppos but i need to use this one so any help is appreciated

**Comment by silver-zepp** [comment, marked as answer]
Our watches don't have node built-in, so it's impossible to run packages that weren't built for zos from the ground up. You have to find some alternative built in vanilla JS in order for it to work; and that alternative should be able to run in a low-ram environment, compared to `xml2js`. 

The best bet would be to offload this to your smartphone. Run a node server with xml2js on it and communicate with it using [ZML](https://www.npmjs.com/package/@zeppos/zml).

**Reply by Noxde** [reply]
Yea, you are right, I didn't even think about offloading it to my phone instead that is a good idea, I'd have to look into it. Thanks for the quick answer I will look into both alternatives

**Reply by Noxde** [reply]
Hi, I was just checking out ZML but I see that it requires the API level 3 to work. I'm using a gtr3 watch, isn't that watch only API level 1? I might be wrong but I wanted to ask just in case. Thanks in advance

**Reply by silver-zepp** [reply]
Sure, it actually has 1.0 support and it should suffice for the task. 
![image](https://github.com/zepp-health/zeppos-forum/assets/138894963/3ac765f8-09a7-4714-9c91-ab570fed97f0)

You can also dig through its code here: https://github.com/zepp-health/zml

As well as trying the [Fetch API](https://github.com/zepp-health/zeppos-samples/tree/620134bfdeac3443f9e65510faaa15970a6e641d/application/1.0/fetch-api) example, which doesn't rely on ZML and uses message builder instead (a predecessor of ZML).

Additional Docs: 
https://docs.zepp.com/docs/samples/app/fetchAPI
https://docs.zepp.com/docs/1.0/guides/best-practice/bluetooth-communication

**Reply by Noxde** [reply]
Thanks once again for the quick response. I'm gonna look into the bluetooth communication docs as well as the ZML repo
